Dev

Haremun opened the pull request:
Please review

Haremun committed:
Added estimates

Haremun committed:
Printing board
Created user interface and classes for board
-tests are mising
-todo and estimates updated
-board is printed with toString method
-TreeSet is used to store added fields
-coordinates are stored in separeted class
-class Field stores its coordinates and symbol

Haremun committed:
Creating tests
Created tests for coordinates and adding fields
To do:
-more tests for existing code
-you can add field out of board bounds
-refactor board class
-create game setup class?

Haremun committed:
lost files for previous commit

hongshui commented on src/main/java/com/bieganski/ox/Coordinates.java:
```diff
@@ -0,0 +1,42 @@
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Coordinates that = (Coordinates) o;
+        System.out.println(that);
```
Please delete Czesio

LIttleAncientForestKami replied:
Nooooo! Czesioooo! ;-)

hongshui commented on src/main/java/com/bieganski/ox/Board.java:
```diff
@@ -0,0 +1,49 @@
+    private TreeSet<Field> fields;
+
+    Board(int size) {
+        fields = new TreeSet<>();
```
Consider initializing fields next to variable declaration (private TreeSet<Field> fields = new TreeSet<>()).

hongshui commented on src/main/java/com/bieganski/ox/Board.java:
```diff
@@ -0,0 +1,49 @@
+        StringBuilder stringBuilder = new StringBuilder();
+        Field field = fields.first();
+        for (int y = 0; y < size; y++) {
+            for (int x = 0; x < size; x++) {
```
For in for. Tomek does not like that. You could hide it by moving second loop to separate method, e.g. "addRow(int column)".
Also, you could also use 'row' and 'column' instead of 'x' and 'y'

hongshui commented on src/main/java/com/bieganski/ox/Board.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @Override
+    public String toString() {
```
It took me a while to understand how this method is working. Maybe you could simplify it a bit?

hongshui commented on src/main/java/com/bieganski/ox/Board.java:
```diff
@@ -0,0 +1,49 @@
+        builder.append("]");
+    }
+
+    private boolean checkIsFiledOnGivenCoordinates(Field field, Coordinates coordinates) {
```
Name of this method is quite confusing. Could you change it  for something else?

hongshui commented on src/main/java/com/bieganski/ox/Field.java:
```diff
@@ -0,0 +1,43 @@
+    public int compareTo(Field field) {
+        return field.coordinates.compareTo(this.coordinates);
+    }
+
+
```
Empty spaces

hongshui commented on src/main/java/com/bieganski/ox/Field.java:
```diff
@@ -0,0 +1,43 @@
+
+import java.util.Objects;
+
+public class Field implements Comparable<Field> {
```
public class

LIttleAncientForestKami replied:
Public. No Javadoc.

hongshui commented on src/main/java/com/bieganski/ox/Field.java:
```diff
@@ -0,0 +1,43 @@
+    private Coordinates coordinates;
+    private Symbol symbol;
+
+    public Field(Coordinates coordinates, Symbol symbol) {
```
public constructor

hongshui commented on src/main/java/com/bieganski/ox/ConsoleInterface.java:
```diff
@@ -0,0 +1,20 @@
+import java.io.PrintStream;
+import java.util.Scanner;
+
+public class ConsoleInterface implements UserInterface {
```
public class

hongshui commented on src/main/java/com/bieganski/ox/ConsoleInterface.java:
```diff
@@ -0,0 +1,20 @@
+    private Scanner scanner;
+    private PrintStream out;
+
+    public ConsoleInterface(InputStream inputStream, PrintStream printStream) {
```
public constructor
Also you could provide in constructor Scanner instead of InputStream. This way you could avoid having to use 'new' keyword.

hongshui commented on src/main/java/com/bieganski/ox/Board.java:
```diff
@@ -0,0 +1,49 @@
+        this.size = size;
+    }
+
+    void addFiled(Field field) throws FieldIsNotEmptyException {
```
Where will you handle this exception (so the game would not stop just because someone wanted to add number to field with value)?
Also, you could consider, instead of throwing exception, to take Map class approach: returning false when value cannot be added.

LIttleAncientForestKami replied:
add field unless it's not empty... sounds strange. Please change appropriately. Naming is hard, but important after all.

hongshui commented on src/main/java/com/bieganski/ox/Board.java:
```diff
@@ -0,0 +1,49 @@
+        Field field = fields.first();
+        for (int y = 0; y < size; y++) {
+            for (int x = 0; x < size; x++) {
+                if (checkIsFiledOnGivenCoordinates(field, new Coordinates(x + 1, y + 1))) {
```
When you use +1, it is a bit confusing and can cause problem later, in other places. Please think about some other solution.

LIttleAncientForestKami replied:
To be filed means to be written or recorded or put in records. So, if I am writing in some system, that today at that hour I have done this code review, that is filed. That usually is about putting things "in files". So, it's a peculiar choice of words here as well.

hongshui commented on src/main/java/com/bieganski/ox/Board.java:
```diff
@@ -0,0 +1,49 @@
+    private boolean checkIsFiledOnGivenCoordinates(Field field, Coordinates coordinates) {
+        if (field == null)
+            return false;
+        return field.isOnGivenCoordinates(coordinates);
```
Can be one line if you use ? :

hongshui commented on src/main/java/com/bieganski/ox/Board.java:
```diff
@@ -0,0 +1,49 @@
+        builder.append("[");
+        builder.append(field);
+        builder.append("]");
+    }
```
Can be one line if you use String.format

hongshui commented on src/main/java/com/bieganski/ox/Board.java:
```diff
@@ -0,0 +1,49 @@
+                    field = fields.higher(field);
+                } else {
+                    stringBuilder.append("[ ]");
+                }
```
Also, instead of adding [x] when there is value and [ ] when there is not, you could read value from field (if field is empty aka if checkIsFieldOnGivenCoordinates == false return local final String EMPTY = " ")
So you could replace this condition like that:
* String value =  getValueFromCoordinate(x,y); // declare new Coordinate inside method
* stringBuilder.apppend("[%s]", value);
* if (!value.equals(String.EMPTY)) field = fields.higher(field);

hongshui commented on src/main/java/com/bieganski/ox/Coordinates.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.Objects;
+
+public class Coordinates implements Comparable<Coordinates> {
+    private int x;
```
Same as before - you could use 'column' and 'row' instead of 'x' and 'y'.

hongshui commented on src/main/java/com/bieganski/ox/Coordinates.java:
```diff
@@ -0,0 +1,42 @@
+        int compareY = Integer.compare(coordinates.y, y);
+        return compareY == 0 ? Integer.compare(coordinates.x, x) : compareY;
+    }
+
```
I remember there is some method Comparator.comparing().thenComparing() or something. I do not remember exactly how it works but if you are interested you can check if it can be applied here

hongshui commented on src/main/java/com/bieganski/ox/Coordinates.java:
```diff
@@ -0,0 +1,42 @@
+                "x=" + x +
+                ", y=" + y +
+                '}';
+    }
```
I believe using String.format() could make this function more readable

hongshui commented on src/main/java/com/bieganski/ox/Field.java:
```diff
@@ -0,0 +1,43 @@
+        this.coordinates = coordinates;
+        this.symbol = symbol;
+    }
+
+
```
To many spaces

hongshui commented on src/main/java/com/bieganski/ox/Symbol.java:
```diff
@@ -0,0 +1,5 @@
+package com.bieganski.ox;
+
+public enum Symbol {
```
public enum

hongshui commented on src/main/java/com/bieganski/ox/UserInterface.java:
```diff
@@ -0,0 +1,6 @@
+package com.bieganski.ox;
+
+public interface UserInterface {
```
public interface

hongshui commented on src/test/java/com/bieganski/ox/AppTest.java:
```diff
@@ -3,6 +3,6 @@
 
 @Test
-public class AppTest{
+public class AppTest {
```
Remember to delete this class

LIttleAncientForestKami replied:
Why would you leave empty test in?

hongshui commented on src/test/java/com/bieganski/ox/BoardTest.java:
```diff
@@ -0,0 +1,25 @@
+@Test
+public class BoardTest {
+    @Test()
+    public void testAddingExistingField() {
```
You could you data provided here to check more than one case

hongshui commented on src/test/java/com/bieganski/ox/BoardTest.java:
```diff
@@ -0,0 +1,25 @@
+import static org.testng.Assert.assertEquals;
+
+@Test
+public class BoardTest {
```
You should also test for situations where someone tries to create board or add values to the board outside it's boundries (aka. argument is -1 or 100000000000)

hongshui commented on src/test/java/com/bieganski/ox/FieldTest.java:
```diff
@@ -0,0 +1,31 @@
+@Test
+public class FieldTest {
+    @Test
+    public void testFieldIsOnGivenCoordinates(){
```
Data provider

LIttleAncientForestKami commented on estimates.adoc:
```diff
@@ -0,0 +1,13 @@
+|Creating project, adding to remote repository with collaborators + CRC cards + flow diagram
+|+ Drawing board with given parameters
+|+ Player input validation + Asking players who starts
+|19.12.2019
```
Try planning for whole week and then you'll see earlier how far are you getting and what stops you. That will help with re-estimation.

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/App.java:
```diff
@@ -2,6 +2,18 @@
 
 public class App{
 	public static void main(String[] args){
-		System.out.println("Hello world!");
```
Sensible cut. ;-)

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/App.java:
```diff
@@ -2,6 +2,18 @@
+		} catch (FieldIsNotEmptyException e) {
+			e.printStackTrace();
+		}
+		userInterface.print(board);
```
I do assume magic numbers here will be replaced later on? Or are there some particular choices behind them?

Also - how do you plan on doing automatic play happen?

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/App.java:
```diff
@@ -2,6 +2,18 @@
+			board.addFiled(new Field(new Coordinates(2, 1), Symbol.X));
+			board.addFiled(new Field(new Coordinates(3, 5), Symbol.O));
+		} catch (FieldIsNotEmptyException e) {
+			e.printStackTrace();
```
Less then stellar... why not replace with something better?

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/App.java:
```diff
@@ -2,6 +2,18 @@
+			board.addFiled(new Field(new Coordinates(2, 2), Symbol.X));
+			board.addFiled(new Field(new Coordinates(6, 7), Symbol.O));
+			board.addFiled(new Field(new Coordinates(2, 1), Symbol.X));
+			board.addFiled(new Field(new Coordinates(3, 5), Symbol.O));
```
Field or Filled was meant here? That is a bit peculiar, that you are adding Fields to the Board. Does it need to grow in some way DURING the game? Or is it filling the field perhaps or occupying it?

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/Board.java:
```diff
@@ -0,0 +1,49 @@
+
+class Board {
+    private int size;
+    private TreeSet<Field> fields;
```
An interesting choice. :+1:

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/Board.java:
```diff
@@ -0,0 +1,49 @@
+
+    void addFiled(Field field) throws FieldIsNotEmptyException {
+        if (fields.contains(field))
+            throw new FieldIsNotEmptyException("You can't add existing field");
```
Not empty or not existing? A bit confusing.

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/Board.java:
```diff
@@ -0,0 +1,49 @@
+
+import java.util.TreeSet;
+
+class Board {
```
Notice please, that you have 2 responsibilities here. There are 2 (two) reasons for you to change that class. When printing of your board should happen and when you will be changing the way of filling the fields out. That's not that great, from SRP perspective.

hongshui replied:
Hi. Yesterday I had method generating board as String in BoardPrinter class but you have told me that something like that should be moved to toString() method in Board class. Now you are telling us that having it in Board class is breaking SRP, so we are both a bit confused. Could you please explain?

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/ConsoleInterface.java:
```diff
@@ -0,0 +1,20 @@
+    private PrintStream out;
+
+    public ConsoleInterface(InputStream inputStream, PrintStream printStream) {
+        this.scanner = new Scanner(inputStream);
```
You can't replace scanner now with something else if you will need it.
I am looking forward to tests of your console interface, with that said. :-)

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/Coordinates.java:
```diff
@@ -0,0 +1,42 @@
+
+import java.util.Objects;
+
+public class Coordinates implements Comparable<Coordinates> {
```
Public yet without Javadocs. Can't go in.

Also - why do you need coordinates?

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/Coordinates.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
So, `Objects.hash` but no `Objects.equals`?

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/Coordinates.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public int compareTo(Coordinates coordinates) {
+        int compareY = Integer.compare(coordinates.y, y);
```
Nifty.

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/Field.java:
```diff
@@ -0,0 +1,43 @@
+
+
+    boolean isOnGivenCoordinates(Coordinates coordinates) {
+        return coordinates.equals(this.coordinates);
```
Shouldn't this delegate to local equals method?

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/FieldIsNotEmptyException.java:
```diff
@@ -0,0 +1,7 @@
+package com.bieganski.ox;
+
+public class FieldIsNotEmptyException extends Exception {
```
D'Exceptions do deserve documentation - DDDD. :P

LIttleAncientForestKami commented on src/test/java/com/bieganski/ox/BoardTest.java:
```diff
@@ -0,0 +1,25 @@
+
+@Test
+public class BoardTest {
+    @Test()
```
No need for parens.
No need for the annotation actually, since there's one on class.
Magic numbers in the test - shall it only works for Board(10), Coords(5,5) and when X is inputted? And if the board is larger or smaller, or char is an O?
Before return: assert on message being as you expect it.

LIttleAncientForestKami commented on src/test/java/com/bieganski/ox/ConsoleInterfaceTest.java:
```diff
@@ -0,0 +1,10 @@
+import java.io.PrintStream;
+
+@Test
+public class ConsoleInterfaceTest {
```
And here I was looking forward to it :( Give me back my expectations!

LIttleAncientForestKami commented on src/test/java/com/bieganski/ox/FieldTest.java:
```diff
@@ -0,0 +1,31 @@
+
+@Test
+public class FieldTest {
+    @Test
```
No need since class has it.

LIttleAncientForestKami commented on src/test/java/com/bieganski/ox/FieldTest.java:
```diff
@@ -0,0 +1,31 @@
+        //Then
+        assertFalse(isOnGivenCoordinates);
+    }
+
```
Both tests should be parameterized.

LIttleAncientForestKami commented on todo.lst:
```diff
@@ -0,0 +1,16 @@
+-------------filed is not on given coordinates
+-------------field is on given coordinates
+adding existing filed should throw exception
+
```
Previous three lines and their spelling had me smile. :)

LIttleAncientForestKami reviewed (changes requested):
More tests and JDocs please.

Haremun committed:
Working on fixes

Haremun committed:
Board now returns boolen on adding element

Haremun committed:
Removed coordinates
Removed redutant coordinates class
This class was replaced with 'position' field in Field class.

Haremun committed:
Review fixes
Fix mistakes that were pointed in review

Haremun committed:
Test for showing message

hongshui commented on src/test/java/com/bieganski/ox/BoardPrinterTest.java:
```diff
@@ -0,0 +1,42 @@
+        boardPrinter.print(fields, boardSize);
+        //Then
+        assertEquals(actualBoard.toString(), expectedBoard);
+    }
```
This test does not pass. Fix it!

What are you actually testing here? You set up both actualBoard and expectedBoard in lines 16-25. What are lines 26-37 for?

hongshui commented on pom.xml:
```diff
@@ -1,71 +1,89 @@
+        <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
+        <project.java.version>11</project.java.version>
+        <maven.compiler.source>${project.java.version}</maven.compiler.source>
+        <maven.compiler.target>${project.java.version}</maven.compiler.target>
```
You could replace those two with <maven.compiler.release>

hongshui commented on pom.xml:
```diff
@@ -1,71 +1,89 @@
+        <maven.compiler.encoding>${project.build.sourceEncoding}</maven.compiler.encoding>
+        <maven.resources.encoding>${project.build.sourceEncoding}</maven.resources.encoding>
+        <version.plugin.maven.jacoco>0.8.5</version.plugin.maven.jacoco>
+        <version.maven>3.5.3</version.maven>
```
Current maven version is 3.6.3

hongshui commented on src/main/java/com/bieganski/ox/Board.java:
```diff
@@ -0,0 +1,36 @@
+        this.occupiedFields = new TreeSet<>();
+        this.size = size;
+        this.boardPrinter = boardPrinter;
+        this.boardPrinter.print(occupiedFields, size);
```
What is [  this.boardPrinter.print(occupiedFields, size); ] line for? I am confused.... Is this Czesio or something else?

hongshui commented on src/main/java/com/bieganski/ox/Board.java:
```diff
@@ -0,0 +1,36 @@
+    Board(int size) {
+        this.occupiedFields = new TreeSet<>();
+        this.size = size;
+        this.boardPrinter = null;
```
You could replace this with what is called "konstructor teleskopowy". 
So it would be just [ this(size, null) ] instead of three line you have right now

hongshui commented on src/main/java/com/bieganski/ox/Board.java:
```diff
@@ -0,0 +1,36 @@
+            return false;
+        occupiedFields.add(field);
+        if (boardPrinter != null)
+            boardPrinter.print(occupiedFields, size);
```
Ok, I think I've got now what was boardPrinter.print in constructor for. But as you can see, current solution may not be so clear to others. Could think about something different to achieve your goal? The first thing that comes to my mind is observer pattern though if you find some better solution, go for it instead :)

As for now I would suggest at least to move this two lines (checking if printer is present and printing) to separate method. This way you could just call it from here and also from constructor.

hongshui commented on src/main/java/com/bieganski/ox/BoardPrinter.java:
```diff
@@ -0,0 +1,33 @@
+        this.userInterface = userInterface;
+    }
+
+    void print(TreeSet<Field> fieldsWithValues, int boardSide) {
```
This is rather long method. Maybe try to split it into smaller parts?

hongshui commented on src/main/java/com/bieganski/ox/BoardPrinter.java:
```diff
@@ -0,0 +1,33 @@
+                stringBuilder.append("\n");
+        }
+
+        userInterface.refreshAndPrint(stringBuilder.toString());
```
I'm not sure if I would have call for user interface here. For now it means that you have two objects doing the same thing (showing user something) and also you have chain of calls (Board -> BoardPrinter -> UserInterface) every new value is added. Maybe it would be worth to try to implement it somehow different? For example you could try to, after adding new value to board, returning string with board's current shape. Or, if you'd use observer, you could see if sending this new shape string together with notification could make a difference.

hongshui commented on src/main/java/com/bieganski/ox/ConsoleInterface.java:
```diff
@@ -0,0 +1,25 @@
+    }
+
+    @Override
+    public void print(Object object) {
```
Since you are calling "println" method of PrintStream here, I think it would be clearer to name method "println" instead of "print".

hongshui commented on src/main/java/com/bieganski/ox/ConsoleInterface.java:
```diff
@@ -0,0 +1,25 @@
+
+    @Override
+    public void refreshAndPrint(Object object) {
+        out.print("\033[2J\033[1;1H");
```
Does this line work correctly for all PrintStream implementations or only for System.out? I think it is worth checking if you're also not sure.

hongshui commented on src/main/java/com/bieganski/ox/Field.java:
```diff
@@ -0,0 +1,47 @@
+        return String.valueOf(symbol);
+    }
+
+    public boolean isInBounds(int bounds) {
```
public

hongshui commented on src/main/java/com/bieganski/ox/BoardPrinter.java:
```diff
@@ -0,0 +1,33 @@
+
+import java.util.TreeSet;
+
+public class BoardPrinter {
```
public class

hongshui commented on src/main/java/com/bieganski/ox/BoardPrinter.java:
```diff
@@ -0,0 +1,33 @@
+public class BoardPrinter {
+    private UserInterface userInterface;
+
+    public BoardPrinter(UserInterface userInterface) {
```
public

hongshui commented on src/test/java/com/bieganski/ox/BoardTest.java:
```diff
@@ -0,0 +1,88 @@
+
+public class BoardTest {
+
+    @DataProvider
+    public Object[] positiveInteger() {
+        return IntStream.rangeClosed(0, 5000).boxed().toArray();
+    }
+
+    @DataProvider
+    public Object[] negativeInteger() {
+        return IntStream.range(-5000, 0).boxed().toArray();
+    }
```
Nice to see them :)

hongshui reviewed (changes requested):
Requested changes: 
- make sure all your tests are working!
- delete all unnecessary publics
- I would also delete empty test classes (no point of having them right now)

As for all the other comments - they are only suggestions, so I leave decision of what you should do with them (and when) to you.

Haremun committed:
Review fixes #2

Haremun committed:
BoardPainter and Listeners
Added interface BoardListner to observe Board
Ui now implements Board listeners and print board with BoardPinter
class.
BoardPainter now return string. Now Board Painter has one resposibility
and is easier to test.

Haremun committed:
Code cleaning
